File: An 2/Sem 1/MAP/Lab2/src/lab2/Stack.java
package lab2;

/**
 *
 * @author mihai
 */
public class Stack {
    private Student[] elements;
    private int capacity;
    private int size = 0;

    /**
     * 
     * Constructor.
     * 
     * @param capacity The maximum capacity of the stack. Positive integer.
     */
    public Stack(int capacity){
        elements = new Student[capacity];
        this.capacity = capacity;
    }
    
    /**
     * 
     * Pushes a student on top of the stack.
     * If the stack reached its capacity, then allocate more space.
     * 
     * @param student The student to be added to the stack.
     */
    public void push(Student student){
        if (student == null) return;
        if (this.size == this.capacity) {
            Student[] temp = new Student[this.capacity];
            System.arraycopy(this.elements, 0, temp, 0, this.capacity);
            this.capacity *= 2;
            this.elements = new Student[this.capacity];
            System.arraycopy(temp, 0, this.elements, 0, this.capacity / 2);
        }
        this.elements[this.size] = student;
        this.size++;
    }
    
    /**
     * 
     * Returns the topmost element from the stack.
     * 
     * @return 
     */
    public Student pop(){
        if (this.size == 0) return null;
        this.size--;
        elements[this.size + 1] = null;
        return elements[this.size];
    }
    
    /**
     * 
     * Returns the size of the stack.
     * 
     * @return The size of the stack. Positive int.
     */
    public int getSize(){
        return this.size;
    }
    
    /**
     * 
     * Checks if the stack is empty.
     * 
     * @return true if the stack is empty, false otherwise.
     */
    public boolean isEmpty() {
        return (this.size == 0);
    }
    
    /**
     * 
     * Returns a copy of this stack.
     * 
     * @return A copy of this stack.
     */
    public Stack copy() {
        Stack copy = new Stack(this.size);
        System.arraycopy(this.elements, 0, copy.elements, 0, this.size);
        copy.size = this.size;
        
        return copy;
    }
}

File: An 2/Sem 1/MAP/Lab6/C#/Lab2/Repository/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Lab2.Lab2_Model;
using Lab2.Lab2_Utils;
using Lab2.Lab2_Utils.Lab2_LinkedList;

namespace Lab2
{
	namespace Lab2_Repository {
	    class Stack<T> {
			private Lab2.Lab2_Utils.Lab2_LinkedList.LinkedList<T> elements;
			private int size = 0;

			/**
		     * 
		     * Constructor.
		     * 
		     */
			public Stack(){
				elements = new Lab2.Lab2_Utils.Lab2_LinkedList.LinkedList<T>();
			}

	        /**
	         * 
	         * Pushes a student on top of the stack.
	         * If the stack reached its capacity, then allocate more space.
	         * 
	         * @param student The student to be added to the stack.
	         */
			public void push(T data) {
				Node<T> lastNode = this.elements.getLastElement();
				Node<T> nodeToAdd = new Node<T>(data);

				if (lastNode != null) {
					lastNode.next = nodeToAdd;
				} else {
					this.elements.firstNode = nodeToAdd;
				}
				size++;
			}

	        /**
	         * 
	         * Returns the topmost element from the stack.
	         * 
	         * @return 
	         */
	        public T pop()
	        {
				if (this.size == 0) {
					throw new StackException ("Stack is empty!");
				}
				this.size--;
				Node<T> lastNode = this.elements.getLastElement();
				T returnData = lastNode.data;
				this.elements.removeNode(lastNode);
				return returnData;
			}

	        /**
	         * 
	         * Returns the size of the stack.
	         * 
	         * @return The size of the stack. Positive int.
	         */
	        public int getSize()
	        {
	            return this.size;
	        }

	        /**
	         * 
	         * Checks if the stack is empty.
	         * 
	         * @return true if the stack is empty, false otherwise.
	         */
	        public Boolean isEmpty()
	        {
	            return (this.size == 0);
	        }

	        /**
	         * 
	         * Returns a copy of this stack.
	         * 
	         * @return A copy of this stack.
	         */
			public Stack<T> copy() {
				Stack<T> copy = new Stack<T>();
				copy.elements = this.elements.copy();
				copy.size = this.size;

				return copy;
			}
	    }
	}
}
